### PR TITLE
luminous: restarting active ceph-mgr cause glitches in bps and iops metrics

### PR DESCRIPTION
**Pull request by shinobu-x**
http://tracker.ceph.com/issues/22029

### CODE REVIEW

**Comment by shinobu-x** [comment]
```
--- a/src/mon/PGMap.cc
+++ b/src/mon/PGMap.cc
@@@ -1189,26 -1077,29 +1184,48 @@@ void PGMap::apply_incremental(CephConte
      if (t != osd_stat.end()) {
        stat_osd_sub(t->first, t->second);
        osd_stat.erase(t);
 +      osd_epochs.erase(*p);
      }
 +
 +    // remove these old osds from full/nearfull set(s), too
 +    nearfull_osds.erase(*p);
 +    full_osds.erase(*p);
    }
  
++<<<<<<< HEAD
 +  // calculate a delta, and average over the last 2 deltas.
 +  pool_stat_t d = pg_sum;
 +  d.stats.sub(pg_sum_old.stats);
 +  pg_sum_deltas.push_back(make_pair(d, delta_t));
 +  stamp_delta += delta_t;
 +
 +  pg_sum_delta.stats.add(d.stats);
 +  if (pg_sum_deltas.size() > (unsigned)MAX(1, cct ? cct->_conf->mon_stat_smooth_intervals : 1)) {
 +    pg_sum_delta.stats.sub(pg_sum_deltas.front().first.stats);
 +    stamp_delta -= pg_sum_deltas.front().second;
 +    pg_sum_deltas.pop_front();
++=======
+   // skip calculating delta while sum was not synchronized
+   if (!stamp.is_zero() && !pg_sum_old.stats.sum.is_zero()) {
+     utime_t delta_t;
+     delta_t = inc.stamp;
+     delta_t -= stamp;
+     // calculate a delta, and average over the last 2 deltas.
+     pool_stat_t d = pg_sum;
+     d.stats.sub(pg_sum_old.stats);
+     pg_sum_deltas.push_back(make_pair(d, delta_t));
+     stamp_delta += delta_t;
+     pg_sum_delta.stats.add(d.stats);
+     auto smooth_intervals =
+       cct ? cct->_conf->get_val<uint64_t>("mon_stat_smooth_intervals") : 1;
+     if (pg_sum_deltas.size() > smooth_intervals) {
+       pg_sum_delta.stats.sub(pg_sum_deltas.front().first.stats);
+       stamp_delta -= pg_sum_deltas.front().second;
+       pg_sum_deltas.pop_front();
+     }
++>>>>>>> 013a380... mgr: skip first non-zero incremental in PGMap::apply_incremental()
    }
+   stamp = inc.stamp;
  
    update_pool_deltas(cct, inc.stamp, pg_pool_sum_old);
  
@@@ -2250,11 -1901,14 +2267,19 @@@ void PGMap::update_delta
  
    /* Aggregate current delta, and take out the last seen delta (if any) to
     * average it out.
+    * Skip calculating delta while sum was not synchronized.
     */
++<<<<<<< HEAD
 +  result_pool_delta->stats.add(d.stats);
 +  size_t s = MAX(1, cct ? cct->_conf->mon_stat_smooth_intervals : 1);
++=======
+   if(!old_pool_sum.stats.sum.is_zero()) {
+     delta_avg_list->push_back(make_pair(d,delta_t));
+     *result_ts_delta += delta_t;
+     result_pool_delta->stats.add(d.stats);
+   }
+   size_t s = cct ? cct->_conf->get_val<uint64_t>("mon_stat_smooth_intervals") : 1;
++>>>>>>> 013a380... mgr: skip first non-zero incremental in PGMap::apply_incremental()
    if (delta_avg_list->size() > s) {
      result_pool_delta->stats.sub(delta_avg_list->front().first.stats);
      *result_ts_delta -= delta_avg_list->front().second;
```

**Comment by tchaikov** [comment]
this backport is buggy. instead, we should cherry-pick 7a23097c34c35c1bf6ec09e86ed3acbb0807068c also.

File: src/mon/PGMap.cc [hunk 1]
```diff
@@ -1197,18 +1192,26 @@ void PGMap::apply_incremental(CephContext *cct, const Incremental& inc)
...
+    delta_t = inc.stamp;
+    delta_t -= stamp;
+    // calculate a delta, and average over the last 2 deltas.
+    pool_stat_t d = pg_sum;
+    d.stats.sub(pg_sum_old.stats);
+    pg_sum_deltas.push_back(make_pair(d, delta_t));
+    stamp_delta += delta_t;
+    pg_sum_delta.stats.add(d.stats);
+    auto smooth_intervals =
+      cct ? cct->_conf->get_val<uint64_t>("mon_stat_smooth_intervals") : 1;
```
Comment (tchaikov) [inline src/mon/PGMap.cc]: this is buggy.

**Comment by shinobu-x** [comment]
7a23097

File: src/mon/PGMap.cc (hunk 1 above)
Comment (shinobu-x) [reply src/mon/PGMap.cc]: @tchaikov Thanks for your suggestion (7a23097).

**Comment by shinobu-x** [comment]
After cherry-picked 7a23097, there was conflict:
```
--- a/src/common/legacy_config_opts.h
+++ b/src/common/legacy_config_opts.h
@@@ -233,12 -233,7 +232,14 @@@ OPTION(mon_clock_drift_allowed, OPT_FLO
  OPTION(mon_clock_drift_warn_backoff, OPT_FLOAT) // exponential backoff for clock drift warnings
  OPTION(mon_timecheck_interval, OPT_FLOAT) // on leader, timecheck (clock drift check) interval (seconds)
  OPTION(mon_timecheck_skew_interval, OPT_FLOAT) // on leader, timecheck (clock drift check) interval when in presence of a skew (seconds)
- OPTION(mon_pg_stuck_threshold, OPT_INT) // number of seconds after which pgs can be considered stuck inactive, unclean, etc (see doc/control.rst under dump_stuck for more info)
  OPTION(mon_pg_min_inactive, OPT_U64) // the number of PGs which have to be inactive longer than 'mon_pg_stuck_threshold' before health goes into ERR. 0 means disabled, never go into ERR.
++<<<<<<< HEAD
 +OPTION(mon_pg_warn_min_per_osd, OPT_INT)  // min # pgs per (in) osd before we warn the admin
 +OPTION(mon_pg_warn_max_object_skew, OPT_FLOAT) // max skew few average in objects per pg
 +OPTION(mon_pg_warn_min_objects, OPT_INT)  // do not warn below this object #
 +OPTION(mon_pg_warn_min_pool_objects, OPT_INT)  // do not warn on pools below this object #
++=======
++>>>>>>> 7a23097... mon/PGMap: use new-style options
  OPTION(mon_pg_check_down_all_threshold, OPT_FLOAT) // threshold of down osds after which we check all pgs
  OPTION(mon_cache_target_full_warn_ratio, OPT_FLOAT) // position between pool cache_target_full and max where we start warning
  OPTION(mon_osd_full_ratio, OPT_FLOAT) // what % full makes an OSD "full"
@@@ -348,9 -339,6 +348,12 @@@ OPTION(mon_client_ping_timeout, OPT_DOU
  OPTION(mon_client_hunt_interval_backoff, OPT_DOUBLE) // each time we reconnect to a monitor, double our timeout
  OPTION(mon_client_hunt_interval_max_multiple, OPT_DOUBLE) // up to a max of 10*default (30 seconds)
  OPTION(mon_client_max_log_entries_per_message, OPT_INT)
++<<<<<<< HEAD
 +OPTION(mon_max_pool_pg_num, OPT_INT)
 +OPTION(mon_pool_quota_warn_threshold, OPT_INT) // percent of quota at which to issue warnings
 +OPTION(mon_pool_quota_crit_threshold, OPT_INT) // percent of quota at which to issue errors
++=======
++>>>>>>> 7a23097... mon/PGMap: use new-style options
  OPTION(client_cache_size, OPT_INT)
  OPTION(client_cache_mid, OPT_FLOAT)
  OPTION(client_use_random_mds, OPT_BOOL)
diff --cc src/mon/PGMap.cc
index 3bba78f,1fbce57..0000000
--- a/src/mon/PGMap.cc
+++ b/src/mon/PGMap.cc
@@@ -1184,34 -1082,23 +1184,44 @@@ void PGMap::apply_incremental(CephConte
      if (t != osd_stat.end()) {
        stat_osd_sub(t->first, t->second);
        osd_stat.erase(t);
 +      osd_epochs.erase(*p);
      }
 -  }
  
 -  // calculate a delta, and average over the last 2 deltas.
 -  pool_stat_t d = pg_sum;
 -  d.stats.sub(pg_sum_old.stats);
 -  pg_sum_deltas.push_back(make_pair(d, delta_t));
 -  stamp_delta += delta_t;
++<<<<<<< HEAD
 +    // remove these old osds from full/nearfull set(s), too
 +    nearfull_osds.erase(*p);
 +    full_osds.erase(*p);
 +  }
  
 +  // skip calculating delta while sum was not synchronized
 +  if (!stamp.is_zero() && !pg_sum_old.stats.sum.is_zero()) {
 +    utime_t delta_t;
 +    delta_t = inc.stamp;
 +    delta_t -= stamp;
 +    // calculate a delta, and average over the last 2 deltas.
 +    pool_stat_t d = pg_sum;
 +    d.stats.sub(pg_sum_old.stats);
 +    pg_sum_deltas.push_back(make_pair(d, delta_t));
 +    stamp_delta += delta_t;
 +    pg_sum_delta.stats.add(d.stats);
 +    auto smooth_intervals =
 +      cct ? cct->_conf->get_val<uint64_t>("mon_stat_smooth_intervals") : 1;
 +    if (pg_sum_deltas.size() > smooth_intervals) {
 +      pg_sum_delta.stats.sub(pg_sum_deltas.front().first.stats);
 +      stamp_delta -= pg_sum_deltas.front().second;
 +      pg_sum_deltas.pop_front();
 +    }
++=======
+   pg_sum_delta.stats.add(d.stats);
+   auto smooth_intervals =
+     cct ? cct->_conf->get_val<uint64_t>("mon_stat_smooth_intervals") : 1;
+   if (pg_sum_deltas.size() > smooth_intervals) {
+     pg_sum_delta.stats.sub(pg_sum_deltas.front().first.stats);
+     stamp_delta -= pg_sum_deltas.front().second;
+     pg_sum_deltas.pop_front();
++>>>>>>> 7a23097... mon/PGMap: use new-style options
    }
 +  stamp = inc.stamp;
  
    update_pool_deltas(cct, inc.stamp, pg_pool_sum_old);
  
@@@ -2253,13 -1900,10 +2263,17 @@@ void PGMap::update_delta
  
    /* Aggregate current delta, and take out the last seen delta (if any) to
     * average it out.
 +   * Skip calculating delta while sum was not synchronized.
     */
++<<<<<<< HEAD
 +  if(!old_pool_sum.stats.sum.is_zero()) {
 +    delta_avg_list->push_back(make_pair(d,delta_t));
 +    *result_ts_delta += delta_t;
 +    result_pool_delta->stats.add(d.stats);
 +  }
++=======
+   result_pool_delta->stats.add(d.stats);
++>>>>>>> 7a23097... mon/PGMap: use new-style options
    size_t s = cct ? cct->_conf->get_val<uint64_t>("mon_stat_smooth_intervals") : 1;
    if (delta_avg_list->size() > s) {
      result_pool_delta->stats.sub(delta_avg_list->front().first.stats);
```
Only `ours (HEADs)` were selected because I'm only interested in `mon_stat_smooth_intervals`.

However during make, there were errors:
```
[ 17%] Building CXX object src/journal/CMakeFiles/journal.dir/Utils.cc.o
/root/git/ceph/src/mon/PGMap.cc: In member function ‘void PGMap::get_health(CephContext*, const OSDMap&, std::list<std::pair<health_status_t, std::basic_string<char> > >&, std::list<std::pair<health_status_t, std::basic_string<char> > >*) const’:
/root/git/ceph/src/mon/PGMap.cc:3353:46: error: ‘struct md_config_t’ has no member named ‘mon_pg_stuck_threshold’
   utime_t cutoff = now - utime_t(cct->_conf->mon_pg_stuck_threshold, 0);
                                              ^
/root/git/ceph/src/mon/PGMap.cc:3369:18: error: ‘struct md_config_t’ has no member named ‘mon_health_max_detail’
      cct->_conf->mon_health_max_detail, detail);
                  ^
/root/git/ceph/src/mon/PGMap.cc:3377:18: error: ‘struct md_config_t’ has no member named ‘mon_health_max_detail’
      cct->_conf->mon_health_max_detail,  detail);
                  ^
/root/git/ceph/src/mon/PGMap.cc:3385:18: error: ‘struct md_config_t’ has no member named ‘mon_health_max_detail’
      cct->_conf->mon_health_max_detail,  detail);
                  ^
/root/git/ceph/src/mon/PGMap.cc:3393:18: error: ‘struct md_config_t’ has no member named ‘mon_health_max_detail’
      cct->_conf->mon_health_max_detail,  detail);
                  ^
/root/git/ceph/src/mon/PGMap.cc:3402:18: error: ‘struct md_config_t’ has no member named ‘mon_health_max_detail’
      cct->_conf->mon_health_max_detail,  detail);
                  ^
/root/git/ceph/src/mon/PGMap.cc:3418:87: error: ‘struct md_config_t’ has no member named ‘mon_pg_stuck_threshold’
     ss << num_inactive_pgs << " pgs are stuck inactive for more than " << cct->_conf->mon_pg_stuck_threshold << " seconds";
                                                                                       ^
/root/git/ceph/src/mon/PGMap.cc:3430:29: error: ‘struct md_config_t’ has no member named ‘mon_health_max_detail’
       int max = cct->_conf->mon_health_max_detail;
```
To fix them, I ended up with doing manual modifications:
e.g.,
```
From `cct->_conf->mon_pg_stuck_threshold` to `cct->_conf->get_val<int64_t>("mon_pg_stuck_threshold")`
From `cct->_conf->mon_health_max_detail` to `cct->_conf->get_val<uint64_t>("mon_health_max_detail")`
```
In short , changed from `old_style` to `new_style` against configs which ended up with errors.

After all modification, asserted unittest completed including make

```
$ ./bin/unittest_mon_pgmap
[==========] Running 5 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 5 tests from pgmap
[ RUN      ] pgmap.min_last_epoch_clean
[       OK ] pgmap.min_last_epoch_clean (3 ms)
[ RUN      ] pgmap.calc_stats
[       OK ] pgmap.calc_stats (1 ms)
[ RUN      ] pgmap.dump_object_stat_sum_0
[       OK ] pgmap.dump_object_stat_sum_0 (0 ms)
[ RUN      ] pgmap.dump_object_stat_sum_1
[       OK ] pgmap.dump_object_stat_sum_1 (0 ms)
[ RUN      ] pgmap.dump_object_stat_sum_2
[       OK ] pgmap.dump_object_stat_sum_2 (0 ms)
[----------] 5 tests from pgmap (4 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 1 test case ran. (4 ms total)
[  PASSED  ] 5 tests.
```

File: src/common/legacy_config_opts.h [hunk 2]
```diff
@@ -274,7 +272,6 @@ OPTION(mon_health_to_clog, OPT_BOOL)
 OPTION(mon_health_to_clog_interval, OPT_INT)
 OPTION(mon_health_to_clog_tick_interval, OPT_DOUBLE)
 OPTION(mon_health_preluminous_compat, OPT_BOOL)
-OPTION(mon_health_max_detail, OPT_INT) // max detailed pgs to report in health detail
```
Comment (tchaikov) [inline src/common/legacy_config_opts.h]: you dropped some changes in `legacy_config_opts.h` while backporting 7a23097, would you mind putting some notes in the `Conflicts` accordingly?

**Comment by shinobu-x** [comment]
I'm writing up only conflict fixing parts which don't include manual modification from new_style to old_style.

`legacy_config_opts.h`
**Pre Modification**
```
--- a/src/common/legacy_config_opts.h
+++ b/src/common/legacy_config_opts.h
@@@ -233,12 -233,7 +232,14 @@@ OPTION(mon_clock_drift_allowed, OPT_FLO
  OPTION(mon_clock_drift_warn_backoff, OPT_FLOAT) // exponential backoff for clock drift warnings
  OPTION(mon_timecheck_interval, OPT_FLOAT) // on leader, timecheck (clock drift check) interval (seconds)
  OPTION(mon_timecheck_skew_interval, OPT_FLOAT) // on leader, timecheck (clock drift check) interval when in presence of a skew (seconds)
- OPTION(mon_pg_stuck_threshold, OPT_INT) // number of seconds after which pgs can be considered stuck inactive, unclean, etc (see doc/control.rst under dump_stuck for more info)
  OPTION(mon_pg_min_inactive, OPT_U64) // the number of PGs which have to be inactive longer than 'mon_pg_stuck_threshold' before health goes into ERR. 0 means disabled, never go into ERR.
++<<<<<<< HEAD
 +OPTION(mon_pg_warn_min_per_osd, OPT_INT)  // min # pgs per (in) osd before we warn the admin
 +OPTION(mon_pg_warn_max_object_skew, OPT_FLOAT) // max skew few average in objects per pg
 +OPTION(mon_pg_warn_min_objects, OPT_INT)  // do not warn below this object #
 +OPTION(mon_pg_warn_min_pool_objects, OPT_INT)  // do not warn on pools below this object #
++=======
++>>>>>>> 7a23097... mon/PGMap: use new-style options
  OPTION(mon_pg_check_down_all_threshold, OPT_FLOAT) // threshold of down osds after which we check all pgs
  OPTION(mon_cache_target_full_warn_ratio, OPT_FLOAT) // position between pool cache_target_full and max where we start warning
  OPTION(mon_osd_full_ratio, OPT_FLOAT) // what % full makes an OSD "full"
@@@ -348,9 -339,6 +348,12 @@@ OPTION(mon_client_ping_timeout, OPT_DOU
  OPTION(mon_client_hunt_interval_backoff, OPT_DOUBLE) // each time we reconnect to a monitor, double our timeout
  OPTION(mon_client_hunt_interval_max_multiple, OPT_DOUBLE) // up to a max of 10*default (30 seconds)
  OPTION(mon_client_max_log_entries_per_message, OPT_INT)
++<<<<<<< HEAD
 +OPTION(mon_max_pool_pg_num, OPT_INT)
 +OPTION(mon_pool_quota_warn_threshold, OPT_INT) // percent of quota at which to issue warnings
 +OPTION(mon_pool_quota_crit_threshold, OPT_INT) // percent of quota at which to issue errors
++=======
++>>>>>>> 7a23097... mon/PGMap: use new-style options
  OPTION(client_cache_size, OPT_INT)
  OPTION(client_cache_mid, OPT_FLOAT)
  OPTION(client_use_random_mds, OPT_BOOL)
```
**Post Modification**
```
$ diff src/common/legacy_config_opts.h /tmp/legacy_config_opts.h.conflict
235a236
> <<<<<<< HEAD
239a241,242
> =======
> >>>>>>> 7a23097... mon/PGMap: use new-style options
347a351
> <<<<<<< HEAD
350a355,356
> =======
> >>>>>>> 7a23097... mon/PGMap: use new-style options
```
`PGMap.cc`
**Pre Modification**
```
--- a/src/mon/PGMap.cc
+++ b/src/mon/PGMap.cc
@@@ -1184,34 -1082,23 +1184,44 @@@ void PGMap::apply_incremental(CephConte
      if (t != osd_stat.end()) {
        stat_osd_sub(t->first, t->second);
        osd_stat.erase(t);
 +      osd_epochs.erase(*p);
      }
 -  }
  
 -  // calculate a delta, and average over the last 2 deltas.
 -  pool_stat_t d = pg_sum;
 -  d.stats.sub(pg_sum_old.stats);
 -  pg_sum_deltas.push_back(make_pair(d, delta_t));
 -  stamp_delta += delta_t;
++<<<<<<< HEAD
 +    // remove these old osds from full/nearfull set(s), too
 +    nearfull_osds.erase(*p);
 +    full_osds.erase(*p);
 +  }
  
 +  // skip calculating delta while sum was not synchronized
 +  if (!stamp.is_zero() && !pg_sum_old.stats.sum.is_zero()) {
 +    utime_t delta_t;
 +    delta_t = inc.stamp;
 +    delta_t -= stamp;
 +    // calculate a delta, and average over the last 2 deltas.
 +    pool_stat_t d = pg_sum;
 +    d.stats.sub(pg_sum_old.stats);
 +    pg_sum_deltas.push_back(make_pair(d, delta_t));
 +    stamp_delta += delta_t;
 +    pg_sum_delta.stats.add(d.stats);
 +    auto smooth_intervals =
 +      cct ? cct->_conf->get_val<uint64_t>("mon_stat_smooth_intervals") : 1;
 +    if (pg_sum_deltas.size() > smooth_intervals) {
 +      pg_sum_delta.stats.sub(pg_sum_deltas.front().first.stats);
 +      stamp_delta -= pg_sum_deltas.front().second;
 +      pg_sum_deltas.pop_front();
 +    }
++=======
+   pg_sum_delta.stats.add(d.stats);
+   auto smooth_intervals =
+     cct ? cct->_conf->get_val<uint64_t>("mon_stat_smooth_intervals") : 1;
+   if (pg_sum_deltas.size() > smooth_intervals) {
+     pg_sum_delta.stats.sub(pg_sum_deltas.front().first.stats);
+     stamp_delta -= pg_sum_deltas.front().second;
+     pg_sum_deltas.pop_front();
++>>>>>>> 7a23097... mon/PGMap: use new-style options
    }
 +  stamp = inc.stamp;
  
    update_pool_deltas(cct, inc.stamp, pg_pool_sum_old);
  
@@@ -2253,13 -1900,10 +2263,17 @@@ void PGMap::update_delta
  
    /* Aggregate current delta, and take out the last seen delta (if any) to
     * average it out.
 +   * Skip calculating delta while sum was not synchronized.
     */
++<<<<<<< HEAD
 +  if(!old_pool_sum.stats.sum.is_zero()) {
 +    delta_avg_list->push_back(make_pair(d,delta_t));
 +    *result_ts_delta += delta_t;
 +    result_pool_delta->stats.add(d.stats);
 +  }
++=======
+   result_pool_delta->stats.add(d.stats);
++>>>>>>> 7a23097... mon/PGMap: use new-style options
    size_t s = cct ? cct->_conf->get_val<uint64_t>("mon_stat_smooth_intervals") : 1;
    if (delta_avg_list->size() > s) {
      result_pool_delta->stats.sub(delta_avg_list->front().first.stats);
```
**Post Modification**
```
$ diff src/mon/PGMap.cc /tmp/PGMap.cc.conflict
1189a1190
> <<<<<<< HEAD
1212a1214,1222
> =======
>   pg_sum_delta.stats.add(d.stats);
>   auto smooth_intervals =
>     cct ? cct->_conf->get_val<uint64_t>("mon_stat_smooth_intervals") : 1;
>   if (pg_sum_deltas.size() > smooth_intervals) {
>     pg_sum_delta.stats.sub(pg_sum_deltas.front().first.stats);
>     stamp_delta -= pg_sum_deltas.front().second;
>     pg_sum_deltas.pop_front();
> >>>>>>> 7a23097... mon/PGMap: use new-style options
2257a2268
> <<<<<<< HEAD
2262a2274,2276
> =======
>   result_pool_delta->stats.add(d.stats);
> >>>>>>> 7a23097... mon/PGMap: use new-style options
```
I can take exact same procedure to break this PR and make this PR work at anytime.

**Comment by tchaikov** [comment]
@shinobu-x sorry for the confusion. i mean, could you brief the change in the `Conflicts:` section of the commit message? now, it looks like
```
Conflicts:
	src/common/legacy_config_opts.h
	src/mon/PGMap.cc
```

i'd suggest put something like
```
Conflicts:
	src/common/legacy_config_opts.h
	src/mon/PGMap.cc: in 729a089, PGMap::get_health() is removed from master branch, but in luminous, this function is still around for backward compatibility. but that function is still using the old-style options. in this change, i choose to keep some of the old-style options removed in the cherry-picked cleanup to appease PGMap::get_health(), instead of migrating all of them to the new-style because of less efforts.
```

**Comment by shinobu-x** [comment]
@tchaikov Thanks for your kind suggestion.

**Pull request merged**